### ISSUE TITLE
"Responding to Backpressure": preventing recursion

### ISSUE THREAD

**Issue by dhoerl** [mattneub/understandingCombine#4]
Anyone can write a Subscriber interface, and they have two ways of conveying demand to the upstream publisher:
1. send "request(_ demand)"
2. return a demand after receiving a "receive" message

Regardless of how you interpret the second Demand, lets suppose the following:
- Publisher has 1M records
- Operator does something to each record
- Subscriber requests one record at a time

Subscriber, knows  it requested one record, and just got one record, sends the publisher "request(.max(1))", then returns .max(1).

Yikes! Recursion! These receive-request will push down 1M times on the stack (I know, its a contrived case)!

My (proposed) solution to this is to handle the case of a received request message while in the middle of sending receive is to use a flag to note that condition, and when that case happens, to just update the "limit" value. My code is exactly like yours in that the core loop is a while(), which handles the updated limit value just fine, and no recursion.

**Comment by mattneub** [comment]
I agree with all of this, but I'm not going to add anything about it to the "book". The "Custom Subscribers" chapter starts with a disclaimer:

> It doesn't really fulfill all of the obligations of a subscriber, and I haven't given any thought to higher considerations of thread safety.

That covers (with hand-waving) exactly the sort of thing you're talking about; I don't want to complicate the discussion by delving into these sorts of considerations.
